File: backend/internal/app/mysql_metadata.go
package app

import (
	"context"
	"database/sql"
	"errors"
	"regexp"
	"sort"
	"strings"
	"time"

	_ "github.com/go-sql-driver/mysql"
	mysqlcfg "github.com/go-sql-driver/mysql"
)

var datasourceDSNPattern = regexp.MustCompile(`[^\s]+:[^\s@]+@tcp\([^)]+\)/[^\s]*`)

var datasourceConnectionTester = runDatasourceConnectionTest
var datasourceDatabaseLister = listDatasourceDatabases
var datasourceTableLister = listDatasourceTables
var datasourceColumnLister = listDatasourceColumns

func runDatasourceConnectionTest(datasource Datasource) DatasourceTestResult {
	startedAt := time.Now()
	testedAt := now()
	if datasource.IsDemo {
		version := datasource.Version
		if strings.TrimSpace(version) == "" {
			version = "MySQL 8.0.44"
		}
		return DatasourceTestResult{
			Success:   true,
			Status:    DatasourceAvailable,
			Version:   version,
			LatencyMS: latencyMilliseconds(startedAt),
			TestedAt:  testedAt,
			Message:   "Connection available",
		}
	}

	db, err := openMySQL(datasource, datasource.DefaultSchema)
	if err != nil {
		return failedDatasourceTest(startedAt, testedAt, sanitizeDatasourceError(err.Error(), datasource))
	}
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	if err := db.PingContext(ctx); err != nil {
		return failedDatasourceTest(startedAt, testedAt, sanitizeDatasourceError(err.Error(), datasource))
	}
	version := ""
	if datasource.Type == DatasourceTypeMySQL {
		var rawVersion string
		if err := db.QueryRowContext(ctx, "SELECT VERSION()").Scan(&rawVersion); err != nil {
			return failedDatasourceTest(startedAt, testedAt, sanitizeDatasourceError(err.Error(), datasource))
		}
		version = formatDatasourceVersion(datasource.Type, rawVersion)
	}
	return DatasourceTestResult{
		Success:   true,
		Status:    DatasourceAvailable,
		Version:   version,
		LatencyMS: latencyMilliseconds(startedAt),
		TestedAt:  testedAt,
		Message:   "Connection available",
	}
}

func formatDatasourceVersion(datasourceType DatasourceType, rawVersion string) string {
	version := strings.TrimSpace(rawVersion)
	if version == "" {
		return ""
	}
	if datasourceType == DatasourceTypeMySQL && !strings.HasPrefix(strings.ToLower(version), "mysql ") {
		return "MySQL " + version
	}
	return version
}

func failedDatasourceTest(startedAt time.Time, testedAt string, message string) DatasourceTestResult {
	message = strings.TrimSpace(message)
	if message == "" {
		message = "Connection failed"
	}
	return DatasourceTestResult{
		Success:   false,
		Status:    DatasourceFailed,
		LatencyMS: latencyMilliseconds(startedAt),
		TestedAt:  testedAt,
		Message:   message,
	}
}

func latencyMilliseconds(startedAt time.Time) int {
	latency := int(time.Since(startedAt).Milliseconds())
	if latency < 1 {
		return 1
	}
	return latency
}

func sanitizeDatasourceError(message string, datasource Datasource) string {
	sanitized := strings.TrimSpace(message)
	if datasource.PasswordSecret != "" {
		if password, err := decryptText(datasource.PasswordSecret); err == nil && password != "" {
			sanitized = strings.ReplaceAll(sanitized, password, "******")
		}
	}
	sanitized = datasourceDSNPattern.ReplaceAllString(sanitized, "[redacted-dsn]")
	if datasource.PasswordSecret != "" && strings.Contains(strings.ToLower(sanitized), "password") {
		sanitized = strings.ReplaceAll(sanitized, datasource.PasswordSecret, "******")
	}
	if sanitized == "" {
		return "Connection failed"
	}
	return sanitized
}

func openMySQL(datasource Datasource, database string) (*sql.DB, error) {
	password, err := decryptText(datasource.PasswordSecret)
	if err != nil {
		return nil, err
	}
	config := mysqlcfg.NewConfig()
	config.User = datasource.Username
	config.Passwd = password
	config.Net = "tcp"
	config.Addr = datasource.Host + ":" + intToString(datasource.Port)
	config.DBName = database
	config.Timeout = 3 * time.Second
	config.ParseTime = true
	return sql.Open("mysql", config.FormatDSN())
}

func listDatasourceDatabases(datasource Datasource) ([]string, error) {
	if datasource.IsDemo {
		return demoDatasourceDatabases(datasource), nil
	}
	db, err := openMySQL(datasource, "")
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	rows, err := db.QueryContext(ctx, "SHOW DATABASES")
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer rows.Close()

	databases := []string{}
	for rows.Next() {
		database := ""
		if err := rows.Scan(&database); err != nil {
			return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
		}
		database = strings.TrimSpace(database)
		if database == "" || isMySQLSystemDatabase(database) {
			continue
		}
		databases = append(databases, database)
	}
	if err := rows.Err(); err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	return databases, nil
}

func listDatasourceTables(datasource Datasource, database string) ([]string, error) {
	database = strings.TrimSpace(database)
	if database == "" {
		return nil, errors.New("DB 必填")
	}
	if datasource.IsDemo {
		return demoDatasourceTables(database), nil
	}
	db, err := openMySQL(datasource, "")
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	rows, err := db.QueryContext(ctx, `
SELECT table_name
FROM information_schema.tables
WHERE table_schema = ? AND table_type = 'BASE TABLE'
ORDER BY table_name
`, database)
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer rows.Close()

	tables := []string{}
	for rows.Next() {
		table := ""
		if err := rows.Scan(&table); err != nil {
			return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
		}
		table = strings.TrimSpace(table)
		if table != "" {
			tables = append(tables, table)
		}
	}
	if err := rows.Err(); err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	return tables, nil
}

func listDatasourceColumns(datasource Datasource, database string, table string) ([]DatasourceColumn, error) {
	database = strings.TrimSpace(database)
	table = strings.TrimSpace(table)
	if database == "" {
		return nil, errors.New("DB 必填")
	}
	if table == "" {
		return nil, errors.New("表必填")
	}
	if datasource.IsDemo {
		return demoDatasourceColumns(database, table), nil
	}
	db, err := openMySQL(datasource, "")
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	rows, err := db.QueryContext(ctx, `
SELECT column_name, column_type, is_nullable, column_key, column_default
FROM information_schema.columns
WHERE table_schema = ? AND table_name = ?
ORDER BY ordinal_position
`, database, table)
	if err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	defer rows.Close()

	columns := []DatasourceColumn{}
	for rows.Next() {
		column := DatasourceColumn{}
		nullable := ""
		columnKey := ""
		defaultValue := sql.NullString{}
		if err := rows.Scan(&column.Name, &column.Type, &nullable, &columnKey, &defaultValue); err != nil {
			return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
		}
		column.Name = strings.TrimSpace(column.Name)
		if column.Name == "" {
			continue
		}
		column.Nullable = strings.EqualFold(nullable, "YES")
		column.IsPrimaryKey = strings.EqualFold(columnKey, "PRI")
		if defaultValue.Valid {
			column.DefaultValue = defaultValue.String
		}
		columns = append(columns, column)
	}
	if err := rows.Err(); err != nil {
		return nil, errors.New(sanitizeDatasourceError(err.Error(), datasource))
	}
	return columns, nil
}

func isMySQLSystemDatabase(database string) bool {
	switch strings.ToLower(strings.TrimSpace(database)) {
	case "information_schema", "mysql", "performance_schema", "sys":
		return true
	default:
		return false
	}
}

func demoDatasourceDatabases(datasource Datasource) []string {
	if strings.TrimSpace(datasource.DefaultSchema) != "" {
		return []string{strings.TrimSpace(datasource.DefaultSchema)}
	}
	return []string{"canal_plus"}
}

func demoDatasourceTables(database string) []string {
	if strings.TrimSpace(database) == "" {
		return []string{}
	}
	return []string{"orders", "users"}
}

func demoDatasourceColumns(database string, table string) []DatasourceColumn {
	if strings.TrimSpace(database) == "" || strings.TrimSpace(table) == "" {
		return []DatasourceColumn{}
	}
	columnsByTable := map[string][]DatasourceColumn{
		"orders": {
			{Name: "id", Type: "bigint", Nullable: false, IsPrimaryKey: true},
			{Name: "user_id", Type: "bigint", Nullable: false},
			{Name: "amount", Type: "decimal(12,2)", Nullable: false, DefaultValue: "0.00"},
			{Name: "status", Type: "varchar(32)", Nullable: false, DefaultValue: "pending"},
			{Name: "created_at", Type: "datetime", Nullable: false},
		},
		"users": {
			{Name: "id", Type: "bigint", Nullable: false, IsPrimaryKey: true},
			{Name: "name", Type: "varchar(128)", Nullable: false},
			{Name: "email", Type: "varchar(255)", Nullable: true},
			{Name: "updated_at", Type: "datetime", Nullable: true},
		},
	}
	columns := columnsByTable[strings.ToLower(strings.TrimSpace(table))]
	if len(columns) == 0 {
		return []DatasourceColumn{
			{Name: "id", Type: "bigint", Nullable: false, IsPrimaryKey: true},
			{Name: "name", Type: "varchar(128)", Nullable: false},
			{Name: "updated_at", Type: "datetime", Nullable: true},
		}
	}
	copied := append([]DatasourceColumn(nil), columns...)
	sort.SliceStable(copied, func(left, right int) bool {
		if copied[left].IsPrimaryKey != copied[right].IsPrimaryKey {
			return copied[left].IsPrimaryKey
		}
		return false
	})
	return copied
}

File: backend/src/mysqlMetadata.ts
import mysql from "mysql2/promise";
import type { Datasource, TableColumn, TableInfo } from "./types.js";
import { decryptText } from "./security.js";

const demoCatalog: Record<string, Record<string, TableInfo[]>> = {
  order_center: {
    tables: [
      {
        schema: "order_center",
        name: "orders",
        engine: "InnoDB",
        rows: 184260,
        columns: [
          { name: "id", type: "bigint", nullable: false, primaryKey: true },
          { name: "customer_id", type: "bigint", nullable: false, primaryKey: false },
          { name: "status", type: "varchar(32)", nullable: false, primaryKey: false },
          { name: "total_amount", type: "decimal(12,2)", nullable: false, primaryKey: false },
          { name: "updated_at", type: "datetime", nullable: false, primaryKey: false }
        ]
      },
      {
        schema: "order_center",
        name: "payments",
        engine: "InnoDB",
        rows: 42618,
        columns: [
          { name: "id", type: "varchar(32)", nullable: false, primaryKey: true },
          { name: "order_id", type: "bigint", nullable: false, primaryKey: false },
          { name: "amount", type: "decimal(12,2)", nullable: false, primaryKey: false },
          { name: "channel", type: "varchar(24)", nullable: false, primaryKey: false },
          { name: "updated_at", type: "datetime", nullable: false, primaryKey: false }
        ]
      }
    ]
  },
  reporting: {
    tables: [
      {
        schema: "reporting",
        name: "ods_orders",
        engine: "InnoDB",
        rows: 184260,
        columns: [
          { name: "id", type: "bigint", nullable: false, primaryKey: true },
          { name: "customer_id", type: "bigint", nullable: false, primaryKey: false },
          { name: "status", type: "varchar(32)", nullable: false, primaryKey: false },
          { name: "total_amount", type: "decimal(12,2)", nullable: false, primaryKey: false },
          { name: "updated_at", type: "datetime", nullable: false, primaryKey: false }
        ]
      },
      {
        schema: "reporting",
        name: "ods_payments",
        engine: "InnoDB",
        rows: 42618,
        columns: [
          { name: "id", type: "varchar(32)", nullable: false, primaryKey: true },
          { name: "order_id", type: "bigint", nullable: false, primaryKey: false },
          { name: "amount", type: "decimal(12,2)", nullable: false, primaryKey: false },
          { name: "channel", type: "varchar(24)", nullable: false, primaryKey: false },
          { name: "updated_at", type: "datetime", nullable: false, primaryKey: false }
        ]
      }
    ]
  }
};

function connectionConfig(datasource: Datasource, database?: string) {
  return {
    host: datasource.host,
    port: datasource.port,
    user: datasource.username,
    password: decryptText(datasource.passwordSecret),
    database,
    connectTimeout: 3000
  };
}

export async function testDatasource(datasource: Datasource) {
  if (datasource.isDemo) {
    return { ok: true, message: "演示数据源连接正常" };
  }

  const connection = await mysql.createConnection(connectionConfig(datasource, datasource.defaultSchema));
  try {
    await connection.query("SELECT 1");
    return { ok: true, message: "连接成功" };
  } finally {
    await connection.end();
  }
}

export async function listSchemas(datasource: Datasource) {
  if (datasource.isDemo) {
    return Object.keys(demoCatalog);
  }

  const connection = await mysql.createConnection(connectionConfig(datasource));
  try {
    const [rows] = await connection.query(
      "SELECT SCHEMA_NAME AS name FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME NOT IN ('mysql', 'information_schema', 'performance_schema', 'sys') ORDER BY SCHEMA_NAME"
    );
    return (rows as Array<{ name: string }>).map((row) => row.name);
  } finally {
    await connection.end();
  }
}

export async function listTables(datasource: Datasource, schema: string) {
  if (datasource.isDemo) {
    return demoCatalog[schema]?.tables.map(({ columns, ...table }) => table) ?? [];
  }

  const connection = await mysql.createConnection(connectionConfig(datasource, schema));
  try {
    const [rows] = await connection.query(
      `SELECT TABLE_SCHEMA AS \`schema\`, TABLE_NAME AS name, ENGINE AS engine, TABLE_ROWS AS \`rows\`
       FROM INFORMATION_SCHEMA.TABLES
       WHERE TABLE_SCHEMA = ?
       ORDER BY TABLE_NAME`,
      [schema]
    );
    return rows as TableInfo[];
  } finally {
    await connection.end();
  }
}

export async function listColumns(datasource: Datasource, schema: string, table: string) {
  if (datasource.isDemo) {
    return demoCatalog[schema]?.tables.find((item) => item.name === table)?.columns ?? [];
  }

  const connection = await mysql.createConnection(connectionConfig(datasource, schema));
  try {
    const [rows] = await connection.query(
      `SELECT
         COLUMN_NAME AS name,
         COLUMN_TYPE AS type,
         CASE WHEN IS_NULLABLE = 'YES' THEN TRUE ELSE FALSE END AS nullable,
         CASE WHEN COLUMN_KEY = 'PRI' THEN TRUE ELSE FALSE END AS primaryKey,
         COLUMN_DEFAULT AS defaultValue
       FROM INFORMATION_SCHEMA.COLUMNS
       WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
       ORDER BY ORDINAL_POSITION`,
      [schema, table]
    );
    return rows as TableColumn[];
  } finally {
    await connection.end();
  }
}
